src/Common/ZooKeeper/ZooKeeperImpl.cpp

```diff
@@ -419,6 +430,9 @@ void ZooKeeper::connect(
                 }
 
                 connected = true;
+
+                initApiVersion(node.address, connection_timeout, node.secure);
```

Comment:
Initialize it via separate connection can have some tricky cases, it is highly unlikely, but consider the following case: there are two nodes zookeeper and clickhouse keeper, and at first connection it goes to the zookeeper and on another to clickhouse keeper, then this will not work correctly.

And another possible scenario, you have balancer, after which you have different keeper instances that has different versions.

Yes this is very unlikely case, since using balancer for coordinator is not what you should do, but still.

Maybe binary interface for this new command can be added? And that way you can use one connection.

Reply:
> two nodes zookeeper and clickhouse keeper,

We don't recommend this in any case, and even if it happens, I initialise second connection after the main one succeeded, i.e. it's guaranteed to be to the same node.
This is only called once, and if it fails we just have a missed optimization (cannot use some Keeper specific operations like FilteredList).

> And another possible scenario, you have balancer, after which you have different keeper instances that has different versions.

Yes, this sounds a bit problematic...

> Maybe binary interface for this new command can be added?

The problem is that if ZooKeeper doesn't recognise something it breaks the connection, and we should keep support for it. If it was only on Keeper's side then it wouldn't be an issue.

Reply:
>The problem is that if ZooKeeper doesn't recognise something it breaks the connection, and we should keep support for it. If it was only on Keeper's side then it wouldn't be an issue

If ZooKeeper will close the connection than it is fine I guess, since you can detect this pretty easy.